proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+
+When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.
+
+### Metadata
+
+#### Metadata emit and code generation
+
+For metadata encoding compiler will rely on recently added [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135). 
+
+Fixed-size buffers like:
+``` C#
+public partial class C
+{
+    public int buffer1[10];
+    public readonly int buffer2[10];
+}
+```
+will be emitted as  fields of a specially decorated struct type.
+
+Equivalent C# code will be:
+
+``` C#
+public partial class C
+{
+    public Buffer10<int> buffer1;
+    public readonly Buffer10<int> buffer2;
+}
+
+[System.Runtime.CompilerServices.InlineArray(10)]
+public struct Buffer10<T>
+{
+    private T _element0;
+
+    [UnscopedRef]
+    public System.Span<T> AsSpan()
+    {
+        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
+    }
+
+    [UnscopedRef]
+    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
+    {
+        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
+        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
+    }
+}
+```
+
+The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
+a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
+will be "speakable" in order to support consumption from other languages. 
+
+A code generated for an access like: 
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer1[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2[1];
+    }
+}
+```
+
+will be equivalent to:
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer.AsSpan()[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2.AsReadOnlySpan()[1];
+    }
+}
+```
+
+#### Metadata import
+
+When compiler imports a field declaration of type *T* and the following conditions are all met:
+- *T* is a struct type decorated with the ```InlineArray``` attribute, and
+- The first instance field declared within *T* has type *F*, and
+- There is a ```public System.Span<F> AsSpan()``` within *T*, and 
+- There is a ```public readonly System.ReadOnlySpan<T> AsReadOnlySpan()``` or ```public System.ReadOnlySpan<T> AsReadOnlySpan()``` within *T*. 
```

Comment:
If the compiler is not going to validate that these AsSpan/AsReadOnlySpan methods have expected implementation, it opens possibility for some very creative hacks. For example, somebody can provide a custom AsReadOnlySpan implementation that returns variable sized Span. (Just an observation.)

Reply:
> If the compiler is not going to validate that these AsSpan/AsReadOnlySpan methods have expected implementation, it opens possibility for some very creative hacks. For example, somebody can provide a custom AsReadOnlySpan implementation that returns variable sized Span. (Just an observation.)

The validation of method bodies is not proposed. In general, compiler is not concerned about types violating a pattern they pretend to satisfy. There is an option for these helpers to be automatically provided by runtime on all inline array types with robust to hacks implementation. For example, array types have helpers like that to access element references, etc.